### ISSUE TITLE
Exception rethrows in .net 8.0.2 rethrow wrong exception in nested catch blocks with an await in outer catch block

### ISSUE THREAD

**Issue by EklipZgit** [dotnet/roslyn#72177]
### Description

When rethrowing an exception that was caught inside an outer catch block followed by an await, the outer exception is rethrown instead of the one one from the catch block the `throw;` is inside of. This began with .net 8.0.2, and does not reproduce in 8.0.1

### Reproduction Steps

XUnit + Fluent Validators reproduction tests:
```CSharp

    [Fact]
    public async Task Validate_BehaviorOfInnerNreRethrows()
    {
        bool threw = false;

        try
        {
            // Must be an async method in order to repro
            await MethodWithInnerNreRethrowAsync();
        }
        catch (Exception ex)
        {
            threw = true;
            var nre = ex.Should().BeOfType<NullReferenceException>();
        }

        threw.Should().Be(true);
    }

    [Fact]
    public async Task Validate_BehaviorOfOtherInnerExceptionsRethrows()
    {
        bool threw = false;

        try
        {
            // Must be an async method in order to repro
            await MethodWithInnerAltExcRethrowAsync();
        }
        catch (Exception ex)
        {
            threw = true;
            var myExc = ex.Should().BeOfType<MyExc>();
        }

        threw.Should().Be(true);
    }

    private async Task MethodWithInnerNreRethrowAsync()
    {
        try
        {
            // This is not necessary for the repro, however it is necessary to compile
            // errorless when you comment out the final await Task.Delay(1)
            // in order to demonstrate that it no longer reprodues without
            // an await inside the outer catch.
            await Task.Delay(1);

            throw new InvalidOperationException("");
        }
        catch (InvalidOperationException ex)
        {
            var something = ex.Message;

            try
            {
                // Throws NRE
                something = ex.InnerException!.Message;
            }
            catch
            {
                // Should rethrow NRE, rethrows the InvalidOperationException instead, when there is an await in the outer catch.
                throw;
            }

            // This await AnythingAsync() must be here, or the problem no longer reproduces.
            await Task.Delay(1);
        }

        // Moving the await outside of the outer catch to the end of the method like below will also no longer reproduce the bug.
        //await Task.Delay(1);
    }

    /// <summary>
    /// Alt variation of the above that proves that this is not just applicable to NRE exceptions, but other exception types as well.
    /// </summary>
    /// <returns></returns>
    private async Task MethodWithInnerAltExcRethrowAsync()
    {
        var a = 1;

        try
        {
            await Task.Delay(1);

            throw new InvalidOperationException("");
        }
        catch (InvalidOperationException)
        {
            try
            {
                // Trick the compiler into not complaining about unreachable Task.Delay below.
                if (a == 1)
                    throw new MyExc("whatever"); // Throws MyExc
            }
            catch
            {
                // Should rethrow MyExc, rethrows the InvalidOperationException instead.
                throw;
            }

            // This await AnythingAsync() must be here, or the problem no longer reproduces.
            await Task.Delay(1);
        }
    }

    public class MyExc : Exception
    {
        public MyExc(string message)
            : base(message)
        {
        }
    }
```

### Expected behavior

The inner exception should be rethrown in both tests

### Actual behavior

The exception from the outer catch block that contains an await inside of it (but outside of the inner try-catch) is rethrown, instead. This can even be seen with the visual studio debugger which was how I first found the issue before writing this set of repro tests. The inner throw statement rethrows the outer exception.

### Regression?

Yes. We've had a test that relies on this since .net 6, made it all the way through net 7 and the net8 upgrade, and began failing as soon as run in a .net 8.0.2 patched container, and now reproducing locally after updating local to 8.0.2.

Both tests pass on 8.0.100 and both fail on 8.0.200

### Known Workarounds

None, besides removing all awaits from the outer catch block and moving them outside the outer catch block.

**Comment by tboby** [comment]
> This is a C# compiler issue. Presumably it's a duplicate of [dotnet/roslyn#71569](https://github.com/dotnet/roslyn/issues/71569).

Yup, I've now found https://github.com/dotnet/sdk/issues/38856, https://github.com/dotnet/roslyn/issues/72141 and that one.

I think this is still reproducing on `mcr.microsoft.com/dotnet/nightly/sdk:8.0`

**Comment by jjonescz** [comment]
FYI: fix for this has been released in SDK 8.0.203 and VS 17.9.2

Closing as duplicate
